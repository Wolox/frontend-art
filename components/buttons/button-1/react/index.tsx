import React from 'react';

import '../styles.scss';

function Button1() {
  return (
    <div className="column center btn-group-1">
      <button className="btn btn-1 m-bottom-2" disabled={true}>
        <div className="btn__content">cancel</div>
      </button>
      <button type="button" className="btn btn-1">
        <div className="btn__content">submit</div>
      </button>
    </div>
  );
}

export default Button1;
