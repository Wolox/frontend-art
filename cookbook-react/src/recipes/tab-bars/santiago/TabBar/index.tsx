import React, { Fragment, useMemo, useState } from 'react';
import cn from 'classnames';

import styles from './styles.module.scss';

export type OptionId = string | number;

interface Option<Id extends OptionId> {
  title: string;
  disabled?: boolean;
  id: Id;
}

interface Props<Id extends OptionId> {
  options: Option<Id>[];
  onChange: (id: Id) => void;
  className?: string;
}

interface InternallyControlledProps<Id extends OptionId> extends Props<Id> {
  active?: never;
  initialValue?: Id;
}

interface ExternallyControlledProps<Id extends OptionId> extends Props<Id> {
  // Optional value to allow external change of tab. If empty, the component sets the clicked
  // tab as the active one.
  active: Id;
  initialValue?: never;
}

function Selector<Id extends OptionId>({
  className = '',
  active,
  onChange,
  options,
  initialValue
}: InternallyControlledProps<Id> | ExternallyControlledProps<Id>) {
  const [internalActiveTab, setInternalActiveTab] = useState<Id | undefined>(active || initialValue);

  const activeValue = useMemo(() => active || internalActiveTab, [active, internalActiveTab]);

  const handleChange = (id: Id) => {
    if (!active) {
      setInternalActiveTab(id);
    }
    onChange(id);
  };

  const nextValueIsSelected = (index: number) =>
    options.length > index + 1 && options[index + 1].id === activeValue;

  return (
    <div className={cn(styles.optionsContainer, className)}>
      {options?.map((option, index) => (
        <Fragment key={`${option.id}`}>
          <button
            type="button"
            onClick={() => handleChange(option.id)}
            className={cn(styles.option, {
              [styles.disabled]: option.disabled,
              [styles.active]: option.id === activeValue
            })}
            disabled={option.disabled}
          >
            {option.title}
          </button>
          {index < options.length - 1 && (
            <div
              className={cn(styles.separator, {
                [styles.active]: nextValueIsSelected(index)
              })}
            />
          )}
        </Fragment>
      ))}
    </div>
  );
}

export default Selector;
