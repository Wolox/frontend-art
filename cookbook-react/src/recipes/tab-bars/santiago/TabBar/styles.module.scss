$color-gray: #E1E1E1;
$color-black: #000;
$color-white: #FFF;
$color-blue: #003DD5;
$btn-height: 45px;
$options: 3;
$btn-width: (100 / $options) * 1%;

$transition-time: 0.25s;

.options-container {
  display: flex;
}

.separator {
  background-color: $color-gray;
  min-height: 100%;
  min-width: 3px;
  transition: background-color $transition-time ease;

  &.active {
    background-color: $color-blue;
  }
}

.option {
  align-items: center;
  border-top: 3px solid $color-gray;
  border-bottom: 3px solid $color-gray;
  color: $color-black;
  cursor: pointer;
  display: flex;
  font-weight: 500;
  height: $btn-height;
  justify-content: center;
  min-width: 100px;
  transition: border-color $transition-time ease, color $transition-time ease;
  width: $btn-width;

  &:first-child {
    border-left: 3px solid $color-gray;
    border-radius: 5px 0 0 5px;
  }

  &:last-child {
    border-right: 3px solid $color-gray;
    border-radius: 0 5px 5px 0;
  }

  &:only-child {
    border-radius: 5px;
  }

  &.disabled {
    color: $color-gray;
    font-weight: 600;
  }

  &.active {
    border-color: $color-blue;
    color: $color-blue;
    font-weight: 600;
  }

  &.active + .separator {
    background-color: $color-blue;
  }
}
