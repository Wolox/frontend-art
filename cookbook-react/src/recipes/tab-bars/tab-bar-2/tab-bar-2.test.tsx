import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import TabBar from './index';

const tabs = [
  { id: 1, label: 'Primera tab', icon: ':D' },
  { id: 2, label: 'Segunda tab', icon: ':C' }
];
const active = null;
const handleChange = jest.fn();

describe('test tab-bar component', () => {
  test('match to the snapshot', () => {
    const { container } = render(<TabBar tabs={tabs} active={active} handleChange={handleChange} />);
    expect(container).toMatchSnapshot();
  });

  test('call handleChange function with the tab id', () => {
    render(<TabBar tabs={tabs} active={active} handleChange={handleChange} />);
    userEvent.click(screen.getByLabelText('Primera tab'));
    expect(handleChange).toHaveBeenCalledWith(1);
  });

  test('show first tab icon', () => {
    render(<TabBar tabs={tabs} active={active} handleChange={handleChange} />);
    const icon = screen.getByRole('icon-1');
    expect(icon).toBeInTheDocument();
  });

  test('select second tab', () => {
    render(<TabBar tabs={tabs} active={active} handleChange={handleChange} />);
    const tabElement = screen.getByLabelText('Segunda tab') as HTMLInputElement;
    fireEvent.change(tabElement, { target: { value: 2 } });
    expect(tabElement.value).toBe('2');
  });
});
