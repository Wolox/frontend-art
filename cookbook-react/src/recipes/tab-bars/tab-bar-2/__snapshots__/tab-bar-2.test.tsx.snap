// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test tab-bar component match to the snapshot 1`] = `
<div>
  <div
    class="tabsContainer"
  >
    <label
      class="radioLabel"
      for="tab-1"
    >
      <i
        class="icon :D"
        role="icon-1"
      />
      <input
        class="tabInput"
        id="tab-1"
        name="tab"
        type="radio"
        value="1"
      />
      Primera tab
    </label>
    <label
      class="radioLabel"
      for="tab-2"
    >
      <i
        class="icon :C"
        role="icon-2"
      />
      <input
        class="tabInput"
        id="tab-2"
        name="tab"
        type="radio"
        value="2"
      />
      Segunda tab
    </label>
  </div>
</div>
`;
